Csv/src/Akka.Streams.Csv/CsvFormatter.cs

```diff
@@ -0,0 +1,131 @@
+
+        private readonly string _endOfLine;
+        private readonly ByteString _endOfLineBs;
+
+        private readonly CsvQuotingStyle _quotingStyle;
+        private readonly Encoding _encoding;
+
+        public string CharsetName => _encoding.EncodingName;
+
+        public CsvFormatter(char delimiter, char quoteChar, char escapeChar, string endOfLine, CsvQuotingStyle quotingStyle, Encoding encoding = null)
+        {
+            _delimiter = delimiter;
+
+            _quoteChar = quoteChar;
+            _duplicatedQuote = new string(quoteChar, 2);
+
+            _duplicatedEscape = new string(escapeChar, 2);
+
+            _quoteOrEscapechar = new[] {quoteChar, escapeChar};
+            _needQuoteChars = new[] {'\r', '\n', delimiter};
+
+            _endOfLine = endOfLine;
+            _endOfLineBs = ByteString.FromString(_endOfLine);
+
+            _quotingStyle = quotingStyle;
+
+            _encoding = encoding ?? Encoding.UTF8;
+        }
+
+        
+        [SuppressMessage("ReSharper", "PossibleMultipleEnumeration")]
+        public ByteString ToCsv(IEnumerable fields)
+        {
+            if (fields == null || !fields.GetEnumerator().MoveNext())
+            {
+                return _endOfLineBs;
+            }
+
+            return ByteString.FromString(NonEmptyToCsv(fields), _encoding);
+        }
+
+        private string NonEmptyToCsv(IEnumerable fields)
+        {
+            var builder = new StringBuilder();
+
+            void SplitAndDuplicateQuotesAndEscapes(string field, int splitAt)
+            {
+                var lastIndex = 0;
+                var index = splitAt;
+                while (index > -1)
+                {
+                    builder.Append(field.Substring(lastIndex, index - lastIndex));
+                    builder.Append(field[index] == _quoteChar ? _duplicatedQuote : _duplicatedEscape);
+                    lastIndex = index + 1;
+                    index = field.IndexOfAny(_quoteOrEscapechar, lastIndex);
+                }
+                if (lastIndex < field.Length)
+                {
+                    builder.Append(field.Substring(lastIndex));
+                }
+            }
+
+            void Append(string field)
+            {
+                int splitAt;
+                var quoteIt = RequiresQuoteOrSplit(field, out splitAt);
+                if (quoteIt)
+                {
+                    builder.Append(_quoteChar);
+                    if (splitAt != -1)
+                        SplitAndDuplicateQuotesAndEscapes(field, splitAt);
+                    else
+                        builder.Append(field);
+                    builder.Append(_quoteChar);
+                }
+                else
+                {
+                    builder.Append(field);
+                }
+            }
+
+            var iterator = fields.GetEnumerator();
+            var hasNext = iterator.MoveNext();
+            var value = iterator.Current;
```

Comment:
It should be undefined, but it won't fail because we do a bool check on hasNext before we use it.